### ISSUE TITLE
Channels can easily have unobserved exceptions

### ISSUE THREAD

**Issue by BrennanConroy** [dotnet/runtime#80885]
### Description

A user reported an issue where there was an `UnobservedTaskException` in code that used a `Channel`. The problem was that the `Exception` was observed via `Channel.Reader.WaitToReadAsync()`, but the same exception is also set on a `TaskCompletionSource` that is visible via `Channel.Reader.Completion`. So proper code needs to know that it has to observe the exception twice to avoid an `UnobservedTaskException`.

### Reproduction Steps

Original code that was problematic.
```
TaskScheduler.UnobservedTaskException += (sender, ex) =>
{
    Console.WriteLine(ex.Exception);
};
await Func();

GC.Collect();

Console.ReadKey();

static async Task Func()
{
    var channel = Channel.CreateBounded<int>(2);
    channel.Writer.Complete(new Exception());

    try
    {
        while (await channel.Reader.WaitToReadAsync().ConfigureAwait(false)) // throws
        {
            while (channel.Reader.TryRead(out var item))
            {
            }
        }

        // Observe any errors in the completion task
        await channel.Reader.Completion;
        // We tried to be good citizens and observe the Task, but even this seems like a stretch for consumers to know that they should do
    }
    catch { }
}
```

The fix would be to move `await channel.Reader.Completion;` into a finally, and to wrap it in another try catch.

### Expected behavior

Expected no `UnobservedTaskException` if we observed the exception from one of the read APIs.

At a bare minimum the [Channel docs](https://learn.microsoft.com/dotnet/core/extensions/channels) should show using the `Reader.Completion` property properly.

One suggestion was to make `Reader.Completion` lazy, it looks like that might be possible since we set the `_parent._doneWriting` exception property everywhere that we complete the `Reader.Completion` TCS currently. This would make it so you don't need to observe the `Reader.Completion` task unless you already grabbed the property.

### Actual behavior

`UnobservedTaskException` due to not observing `Reader.Completion`.

### Regression?

No.

### Known Workarounds

Double observe the exception by moving `await Reader.Completion` into a `finally` block.

**Comment by theodorzoulias** [comment]
Interesting. The simplest workaround is probably to observe the `Completion.Exception` property in a `finally` block:

```C#
finally { _ = channel.Reader.Completion.Exception; }
```

[Demo](https://dotnetfiddle.net/9cnXL4).

**Comment by stephentoub** [comment]
It's behaving as designed, though I can see how that design could be a little irritating for anyone paying attention to UnobservedTaskExceptions.

> One suggestion was to make Reader.Completion lazy, it looks like that might be possible since we set the _parent._doneWriting exception property everywhere that we complete the Reader.Completion TCS currently. This would make it so you don't need to observe the Reader.Completion task unless you already grabbed the property.

It's possible but I fear it would lead to some non-trivial complexity in the implementation, in particular around additional required synchronization / locking due to accommodating races between accessing the property and the task being completed.  I'm not particularly excited about that complexity and cost, though if someone wanted to prototype it we could see what it really looked like and whether it was reasonable enough.

> At a bare minimum the [Channel docs](https://learn.microsoft.com/dotnet/core/extensions/channels) should show using the Reader.Completion property properly.

Improving the docs is always welcome, if someone wants to submit a PR to the docs repo with a suggestion.

**Comment by theodorzoulias** [comment]
> It's behaving as designed,

Surprisingly the TPL Dataflow has different behavior though. A faulted block with unobserved exception doesn't trigger the `UnobservedTaskException` event.

```C#
BufferBlock<int> block = new();
((IDataflowBlock)block).Fault(new Exception());
```

[Online demo](https://dotnetfiddle.net/3peSZl).

**Comment by stephentoub** [comment]
> Surprisingly the TPL Dataflow has different behavior though.

That's because it happens to access the Task's Exception as an implementation detail.

**Comment by theodorzoulias** [comment]
Indeed, faulted `ActionBlock`s and `WriteOnceBlock`s do trigger the event. Other blocks like the `BufferBlock`, `BatchBlock`, `TransformBlock`, `TransformManyBlock` and `BroadcastBlock` don't trigger the event.

**Comment by davidfowl** [comment]
This feels like a pit of failure in the current API, and while I hate adding complexity, this one feel bad. Can we observe the exception internally to avoid the unobserved exception handler being raised?

**Comment by stephentoub** [comment]
> Can we observe the exception internally to avoid the unobserved exception handler being raised?

This is what i meant by "If we wanted to change this, it wouldn't need to be so complex. We'd just touch Exception wherever we complete the task."

> This feels like a pit of failure in the current API, and while I hate adding complexity, this one feel bad.

There are some exceptions that aren't surfaced in other ways, so in general code should still consume Completion to observe those. If they don't, they're also possibly missing failures. We can selectively suppress the event only in situations where they're exposed in other ways, but for someone not paying attention to the event, it won't matter, and for someone who is, they'll still need to consume Completion to avoid it. We also can't guarantee exceptions surfaced in other ways are observed, since they're typically not Tasks and the returned IVTS instances could just be dropped by the consumer.

In short, the current approach is as it is because it's consistent. We can trade one set of issues for another of desired.

**Comment by davidfowl** [comment]
> We also can't guarantee exceptions surfaced in other ways are observed

Can you show some an example (based on the above)?

**Comment by stephentoub** [comment]
> Can you show some an example (based on the above)?

e.g. if the value task returned from any of the APIs is dropped

e.g. if the channel is completed with an error and no one reads from it again such that it's only surfaced from Completion

**Comment by davidfowl** [comment]
Got it. It sorta makes you want to avoid using exceptions in the channel API because of the extra ceremony required to handle them. I guess you could also argue that the root of the problem is unobserved task exception handler 😄...

**Comment by theodorzoulias** [comment]
> > Can you show some an example (based on the above)?
> 
> e.g. if the value task returned from any of the APIs is dropped
> 
> e.g. if the channel is completed with an error and no one reads from it again such that it's only surfaced from Completion

Regarding these examples, are these errors that are surfaced only by awaiting the `channel.Reader.Completion`, and are not surfaced by awaiting the `channel.Reader.WaitToReadAsync`/`channel.Reader.ReadAllAsync`? Could we have a minimal demo of an example?

**Comment by stephentoub** [comment]
> Could we have a minimal demo of an example?

Just don't call those methods after the channel is completed with an error, or call them without awaiting them. The whole point of the unobserved event is to say "hey, there's an exception you may have neglected to notice due to an issue in how you've constructed your code".

**Comment by theodorzoulias** [comment]
The problem is that the `TaskScheduler.UnobservedTaskException` event is raised even when the `WaitToReadAsync`/`ReadAllAsync` methods are awaited properly:

```C#
TaskScheduler.UnobservedTaskException += (sender, ex) =>
{
    Console.WriteLine($"UnobservedTaskException: {ex.Exception.InnerException.Message}");
};
await DoWork();
GC.Collect();
Thread.Sleep(500);

static async Task DoWork()
{
    Channel<int> channel = Channel.CreateUnbounded<int>();
    channel.Writer.Complete(new Exception());
    try
    {
        await foreach (int item in channel.Reader.ReadAllAsync());
    }
    catch (Exception ex)
    {
        Console.WriteLine($"Await foreach failed: {ex.Message}");
    }
}
```

Output:

```
Await foreach failed: Exception of type 'System.Exception' was thrown.
UnobservedTaskException: Exception of type 'System.Exception' was thrown.
```

[Online demo](https://dotnetfiddle.net/YPnKZl).

Is there any scenario that after consuming the channel with `ReadAllAsync`, someone should also `await` the `Completion`, in order to observe some exception that was not surfaced previously by the `ReadAllAsync`? I am asking in order to clarify this quote:

> There are some exceptions that aren't surfaced in other ways, so in general code should still consume Completion to observe those.

**Comment by stephentoub** [comment]
> The problem is that the TaskScheduler.UnobservedTaskException event is raised even when the WaitToReadAsync/ReadAllAsync methods are awaited properly

I understand that.  I was saying that if we change the implementation to suppress the event, we'd be suppressing the event in some cases where the exception was already seen through another task in exchange for also suppressing the event in cases where the exception was never seen.

**Comment by davidfowl** [comment]
Can we make this one scenario better then? It seems like it would be worthwhile and not add extra complexity. Then we can see if other situations arise (like the scenarios you mentioned) in practice.

**Comment by stephentoub** [comment]
> Can we make this one scenario better then?

Which one exactly?

**Comment by theodorzoulias** [comment]
Imaginary implementation that would fix this problem in a custom channel:

```C#
public sealed class ChannelWrapper<T> : Channel<T>
{
    private readonly Channel<T> _channel;

    private class Reader : ChannelReader<T>
    {
        private readonly ChannelWrapper<T> _parent;

        public override ValueTask<bool> WaitToReadAsync(CancellationToken cancellationToken = default)
        {
            try
            {
                return await _parent._channel.Reader.WaitToReadAsync(cancellationToken);
            }
            catch
            {
                _ = _parent._channel.Reader.Completion.Exception;
                throw;
            }
        }
    }
}
```

Someone could wrap any built-in implementation in a `ChannelWrapper<T>`, after implementing also the other `abstract` and `virtual` members, just to fix this small problem. Which is obviously overkill.

**Comment by davidfowl** [comment]
I agree with you @stephentoub, but we need to retrain people here then. Customers are hooking this event, logging, and seeing errors in their telemetry systems and then reporting bugs.

**Comment by BrennanConroy** [comment]
UnobservedTaskExceptions are still an issue for UI apps after .NET Framework 4.5. And since Channels (and SignalR client which uses Channels and where this issue came from) target netstandard/net462, this is an issue for those apps.

**Comment by stephentoub** [comment]
> UnobservedTaskExceptions are still an issue for UI apps after .NET Framework 4.5.

Can you elaborate?  What's different about UI apps with regards to UnobservedTaskExceptions?

**Comment by BrennanConroy** [comment]
https://learn.microsoft.com/en-us/dotnet/api/system.windows.application.dispatcherunhandledexception?view=netframework-4.8.1#remarks
> By default, Windows Presentation Foundation catches unhandled exceptions, notifies users of the exception from a dialog box (from which they can report the exception), and automatically shuts down an application.

**Comment by stephentoub** [comment]
> https://learn.microsoft.com/en-us/dotnet/api/system.windows.application.dispatcherunhandledexception?view=netframework-4.8.1#remarks
> 
> > By default, Windows Presentation Foundation catches unhandled exceptions, notifies users of the exception from a dialog box (from which they can report the exception), and automatically shuts down an application.

That's "unhandled", not "unobserved".  UnobservedTaskException doesn't show up anywhere in dotnet/wpf.
https://github.com/dotnet/wpf/search?q=UnobservedTaskException

**Comment by BrennanConroy** [comment]
Ahh, I may have been thinking those were the same 😄 
The names are too close 🙈

**Comment by MikeOnTea** [comment]
FWIW, i reported the original unobserved exception case in SignalR. We explicitly exit our application in the unobserved exception handler, because usually those are caused by a mistake from our developers not properly awaiting tasks, and we want to notice and fix those quickly.
While i also found the behaviour of the Channel-API a little surprising regarding exception observation, i also agree the common case should be efficient. (If it stays as it is though a little notice in the documentation would probably be helpful as already mentioned).
As a library consumer it's mostly fine for me as long as i can use the library in question so that there won't be any unobserved exceptions.

**Comment by theodorzoulias** [comment]
Whenever I am using channels in producer-consumer scenarios, I have the dilemma of how to propagate a possible error of the producer. Should I funnel it through the `Channel<T>`, or let it captured in the [`Exception`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.exception) property of the enclosing `Task`? There are pros and cons with each option. This issue helped me resolve this dilemma, by tipping the balance slightly towards the `Task.Exception` option. :-)
